### PR TITLE
feat(eslint-plugin): add new extended rule `no-restricted-imports`

### LINKED ISSUE #2661: [no-restricted-imports] extend to support type-only imports
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

We need to limit only type import some packages sometimes.
For example:
When using package in user project runtimes, Like vetur, ts-loader.

## Rule Details
This rule limit only type import some package within your codebase.

### Options
```typescript
type Options = {
  packages: Array<{ name: string, reason: string }>
};
```

### Example
Incorrect code for `{ packages: [{ name: 'typescript', reason: 'runtime dependency' }] }`
```typescript
import ts from 'typescript'
import * as ts from 'typescript'
import { findPackageConfig } from 'typescript'
```

correct code for `{ packages: [{ name: 'typescript', reason: 'runtime dependency' }] }`
```typescript
import type ts from 'typescript'
import type * as ts from 'typescript'
import type { findPackageConfig } from 'typescript'
```

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Adds new extended rule [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports). The new rule allows to specify new option `allowTypeImport` for path or pattern. See [the doc](https://github.com/typescript-eslint/typescript-eslint/blob/0fa9b63f18611fa10ba8c9fe1b1d55a0f8ea859f/packages/eslint-plugin/docs/rules/no-restricted-imports.md) for more details.

Closes #2661.
This PR does not address #2563.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-restricted-imports.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+This rule adds the following options:
+
+### `allowTypeImports`
+
+(default: `false`)
+
+You can specify this option for a specific path or pattern as follows:
+
+```jsonc
+"no-restricted-imports": ["error", {
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-restricted-imports.md:29]:
```suggestion
"@typescript-eslint/no-restricted-imports": ["error", {
```

File: packages/eslint-plugin/src/rules/no-restricted-imports.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    function isAllowedTypeImportPattern(importSource: string): boolean {
+      return allowedImportTypeMatchers.every(matcher => {
+        return matcher.ignores(importSource);
+      });
+    }
+
+    return {
+      ImportDeclaration(node): void {
+        if (node.importKind === 'type') {
+          const importSource = (node.source.value as string).trim();
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-restricted-imports.ts:155]: nit - right now our parser allows non-string `source.value`'s - so we need to handle this gracefully.

ideally we can just quit early if the value isn't a string.

File: packages/eslint-plugin/src/rules/no-restricted-imports.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        } else {
+          return rules.ImportDeclaration(node);
+        }
+      },
+      ExportNamedDeclaration(node): void {
+        if (node.source?.type !== AST_NODE_TYPES.Literal) {
+          return;
+        }
+        if (node.exportKind === 'type') {
+          const importSource = (node.source.value as string).trim();
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-restricted-imports.ts:171]: ditto

**Review by bradzacher — CHANGES_REQUESTED** [review]
almost there! a few comments
thanks for your work!

**Comment by sosukesuzuki** [comment]
@bradzacher Thank you for review! I've addressed reviewed points!

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for doing this!

**Pull request merged**